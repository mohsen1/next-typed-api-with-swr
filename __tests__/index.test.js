const path = require("path");
const { NextJsTypedApiWithSwrPlugin } = require("..");
const fs = require("fs");

describe("NextJsTypedApiWithSwrPlugin", () => {
  let spy;
  beforeEach(() => {
    spy = jest.spyOn(fs.promises, "writeFile").mockImplementation(() => {});
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("generates correct content with default configurations", async () => {
    const plugin = new NextJsTypedApiWithSwrPlugin();
    let tapFn;
    const compiler = {
      context: path.resolve(__dirname, "__fixtures__"),
      hooks: {
        thisCompilation: {
          tap: (_name, fn) => {
            tapFn = fn;
          },
        },
      },
    };

    plugin.apply(compiler);
    await tapFn();

    expect(spy.mock.calls[0][0]).toContain("__fixtures__/swr-api-types.ts");
    expect(spy.mock.calls[0][1]).toMatchSnapshot();
  });

  it("generates correct content with custom configurations", async () => {
    const plugin = new NextJsTypedApiWithSwrPlugin({
      outputFilePath: "__generated__/api_types.ts",
    });
    let tapFn;
    const compiler = {
      context: path.resolve(__dirname, "__fixtures__"),
      hooks: {
        thisCompilation: {
          tap: (_name, fn) => {
            tapFn = fn;
          },
        },
      },
    };

    plugin.apply(compiler);
    await tapFn();

    expect(spy.mock.calls[0][0]).toContain(
      "__fixtures__/__generated__/api_types.ts"
    );
    expect(spy.mock.calls[0][1]).toMatchSnapshot();
  });
});
