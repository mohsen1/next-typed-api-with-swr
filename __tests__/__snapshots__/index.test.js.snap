// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextJsTypedApiWithSwrPlugin generates correct content with custom configurations 1`] = `
"import { NextApiHandler } from \\"next\\";
import { Fetcher, SWRResponse } from \\"swr\\";
import ___pages_api_users from \\"../pages/api/users\\";


type InferNextApiHandlerResponseType<T extends NextApiHandler> =
  T extends NextApiHandler<infer U> ? U : never;

declare module \\"swr\\" {
  export interface SWRHook {

    <Error = any>(
      key: \\"/api/users\\",
      fetcher?: Fetcher<
        InferNextApiHandlerResponseType<typeof ___pages_api_users>,
        \\"/api/users\\"
      >
    ): SWRResponse<
      InferNextApiHandlerResponseType<typeof ___pages_api_users>,
      Error
    >;
  }
}
"
`;

exports[`NextJsTypedApiWithSwrPlugin generates correct content with default configurations 1`] = `
"import { NextApiHandler } from \\"next\\";
import { Fetcher, SWRResponse } from \\"swr\\";
import pages_api_users from \\"pages/api/users\\";


type InferNextApiHandlerResponseType<T extends NextApiHandler> =
  T extends NextApiHandler<infer U> ? U : never;

declare module \\"swr\\" {
  export interface SWRHook {

    <Error = any>(
      key: \\"/api/users\\",
      fetcher?: Fetcher<
        InferNextApiHandlerResponseType<typeof pages_api_users>,
        \\"/api/users\\"
      >
    ): SWRResponse<
      InferNextApiHandlerResponseType<typeof pages_api_users>,
      Error
    >;
  }
}
"
`;
